### PR TITLE
fix(mcdu): Vapp pilot entry improvements

### PR DESCRIPTION
**Pull request by tracernz**
## Summary of Changes

The pilot can override the Vapp value on PERF APPR page. They can then clear their entered value to revert to the FMS computed value. This fixes a small bug where we cleared the FMS value, but still gave the operator a "NOT ALLOWED" scratchpad message.

## Screenshots (if necessary)

Computed Vapp (small text):
image
Pilot-entered Vapp (normal/large text):
image

## References

Honeywell Step 1A:
image
image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

- Check that pilot entered Vapp can be cleared without a scratchpad message

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -1998,12 +1998,16 @@ class FMCMainDisplay extends BaseAirliners {
...
+                this.vApp = NaN;
+                return true;
+            }
+        } else {
+            const value = parseFloat(s);
+            if (isFinite(value) && value > 0) {
+                this.vApp = value;
+                return true;
+            }
         }
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2011]: maybe want to return entry out of range when it's out of range?

example:
```suggestion
            }
            this.addNewMessage(NXSystemMessage.entryOutOfRange);
            return false;
        }
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 2]
```diff
@@ -1998,12 +1998,18 @@ class FMCMainDisplay extends BaseAirliners {
...
-        if (isFinite(value) && value > 0) {
-            this.vApp = value;
-            return true;
+            if (isFinite(this.vApp)) {
+                this.vApp = NaN;
+                return true;
+            }
+        } else {
+            const value = parseFloat(s);
+            if (isFinite(value) && value > 0) {
```
Comment (beheh) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2007]: The Range is 90 to VMO, so it should be >=90 and <=350.

**Review by beheh — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 2 above)
Comment (tracernz) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js]: Done :)

**Review by davidwalschots — APPROVED** [review]
LGTM

**Review by MisterChocker — APPROVED** [review]
LGTM

**Review by beheh — APPROVED** [review]

**Comment by stescouse** [comment]
Category: QA Tester
Name: stescouse#7271
Date of testing: 26.1.2021
Version of the sim: 1.12.13.0
PR Tested: #3154
Tier of testing: 1
Changes to observe: not allowed on clearing pilot entered Vapp

Testing technique: TryOut
Steps you've made to test the Pull Request:

1. Spawned C&D at EDDF 
2. Took off and flew to EHAM.
3. At APPR entry changed Vapp speed from 133 to 140 observed for warning.

clear appr

clear appr 2

4. Cleared Vapp speed to return to calculated speed.

clear appr 3


Flight notes (If required: EDDF/25L F240 MARUN6N MARUN Y152 ARPEG/F360 Z850 HMM T281 NORKU
 NORKU2A EHAM/27

Issues (If Any): None


Overall Rating: Good
Conclusions of the testing: Vapp speed and warning acted as per PR, no issues.

**Review by tyler58546 — CHANGES_REQUESTED** [review]
Needs changelog

**Review by MisterChocker — CHANGES_REQUESTED** [review]
Automatic calculated value should be with small font, when manually set, normal font.
Do we want to include that in this pr?

**Review by MisterChocker — APPROVED** [review]
This is a small fix, no need to add additional features.

**Comment by tracernz** [comment]
> Automatic calculated value should be with small font, when manually set, normal font.
> Do we want to include that in this pr?

I thought about it but I think it's best done in @beheh's cleanup of the whole page. This was meant to be a tiny bugfix and already got bigger.

**Comment by tracernz** [comment]
> 
> 
> Needs changelog

Done. Thanks for reminder. :)

**Review by tyler58546 — APPROVED** [review]

**Comment by beheh** [comment]
>I thought about it but I think it's best done in @beheh's cleanup of the whole page. This was meant to be a tiny bugfix and already got bigger.

Yeah, I agree (#1824). On the other hand I was planning on deferring the APPR page part of that a bit while I focus on some other projects, so that may take some time. But that should not block this PR.

**Comment by tracernz** [comment]
> Yeah, I agree (#1824). On the other hand I was planning on deferring the APPR page part of that a bit while I focus on some other projects, so that may take some time. But that should not block this PR.

I could pick that up and bring the relevant bits of #1824 into a new PR if you like.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 3]
```diff
@@ -1998,12 +1998,18 @@ class FMCMainDisplay extends BaseAirliners {
...
-        const value = parseFloat(s);
-        if (isFinite(value) && value > 0) {
-            this.vApp = value;
-            return true;
+            if (isFinite(this.vApp)) {
+                this.vApp = NaN;
+                return true;
+            }
+        } else {
+            const value = parseFloat(s);
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:2006]: I am just seeing this we should use parseInt instead! (or detect a decimal point and throw "not allowed")

**Review by MisterChocker — COMMENTED** [review]
I think we should take the time and add this to the pr, since it's an easy fix.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js (hunk 3 above)
Comment (tracernz) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:1985]: The more you look at the existing code... I will fix ;).

**Comment by tracernz** [comment]
> 
> 
> I think we should take the time and add this to the pr, since it's an easy fix.

Okay, I'll make this a draft until done.

**Comment by tracernz** [comment]
Added the small/large text for computed/pilot Vapp, and format error for decimal point as discussed.
Computed Vapp (small text):
image
Pilot-entered Vapp (normal/large text):
image

**Review by MisterChocker — APPROVED** [review]
LGTM

**Review by MisterChocker — APPROVED** [review]
LGTM

**Review by tyler58546 — APPROVED** [review]

**Review by davidwalschots — APPROVED** [review]
LGTM

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 28/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3154 
Tier of testing: 1
Changes to observe: Vapp manual entry improvements

Testing technique: Spawned on approach to LOWS - fiddled with MCDU data
Steps you've made to test the Pull Request:
1.  Spawned inbound LOWS 40nm West
2.  Input all arrival data into MCDU - amended Vapp manually a few times
3. Wiped it a few times to check it returns to auto filled data value
4. Landed

Flight notes (When done a full flight): N/A
Media: N/A

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: All good, no issues found during testing.

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov#8337
Date of testing: 28/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3154
Tier of testing: 1
Changes to observe: Vapp pilot entry improvements

Testing technique: Tryout

Steps you've made to test the Pull Request:

1. Spawned C/D at VABB gate 5. Prepared for a normal IFR flight from VABB to VOML.
2. Observed a normal flight till approach.
3. On approach, planned Flap 3 landing at VOML. Checked APP speeds on MCDU.
4. Cleared the value for Vapp and re-entered new values.
5. Proceeded with the approach normally.

Flight notes (When done a full flight): VABB SID MABTA W15 KANTI DCT VOML
Weather: REX 6

Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: Was able to clear and re-enter values of Vapp on approach, good to go!

**Comment by HugoTro** [comment]
Category: QA Tester
Name: Hugo #3307
Date of testing: Current date 28/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3154 
Tier of testing: 1
Changes to observe: (New) Vapp entry

Testing technique: Try-out
Steps you've made to test the Pull Request:
1. Spawned in C&D and powered the plane up
2. Went into PERF page and then APPR phase
3. Tried different entries and cleared to come back to the computed speed

Issues (If Any): None
Steps to achieve the issue: /

Overall Rating: Good
Conclusions of the testing: Vapp pilot entry is allowed without error when in range, and when there is no slashes or dots, when cleared, it goes back to the smaller font for the computed speed.

**Pull request merged**